automerge-frontend/src/state_tree/optimistic.rs

```diff
@@ -0,0 +1,283 @@
+use std::ops::{Deref, DerefMut};
+
+use super::{MultiGrapheme, MultiValue, ResolvedPathMut, StateTree};
+use crate::{path::PathElement, Path};
+
+/// Contains the required data to undo an operation on the state tree.
+#[derive(Clone, Debug)]
+pub(crate) enum LocalOperationForRollback {
+    Set { old: Option<MultiValue> },
+    SetList { old: MultiValue },
+    SetText { old: MultiGrapheme },
+    Delete { old: MultiValue },
+    DeleteText { old: MultiGrapheme },
+    Insert,
+    InsertMany { count: usize },
+    Increment { by: i64 },
+}
+
+/// Keeps track of the changes made to a state tree and allows rolling back changes.
+#[derive(Clone, Debug)]
+pub(crate) struct OptimisticStateTree {
+    state: StateTree,
+    copies_for_rollback: Vec<(Path, LocalOperationForRollback)>,
+}
+
+impl OptimisticStateTree {
+    pub(crate) fn new(state_tree: StateTree) -> Self {
+        Self {
+            state: state_tree,
+            copies_for_rollback: Vec::new(),
+        }
+    }
+
+    pub(crate) fn take_state(&mut self) -> StateTree {
+        std::mem::take(&mut self.state)
+    }
+
+    /// Commit the operations, making it possible to roll them back.
+    pub(crate) fn commit_operations(&mut self, ops: Vec<(Path, LocalOperationForRollback)>) {
+        self.copies_for_rollback.extend(ops)
+    }
+
+    /// Rollback a list of operations that have been applied.
+    pub(crate) fn rollback_operations(&mut self, ops: Vec<(Path, LocalOperationForRollback)>) {
+        self.rollback(ops.into_iter())
+    }
+
+    /// Rollback all applied operations.
+    pub fn rollback_all(&mut self) {
+        let rollback_ops = std::mem::take(&mut self.copies_for_rollback);
+        self.rollback(rollback_ops.into_iter())
+    }
+
+    /// Undo the operations applied to this document.
+    ///
+    /// This is used in the case of an error to undo the already applied changes.
```

Comment:
This docstring is slightly inaccurate now right? As we also use this in the case of non-local patches?

Reply:
Yep, I've removed this line now.